entity-framework/core/modeling/value-comparers.md

```diff
@@ -0,0 +1,147 @@
+
+However, that being said, it is common to use properties of types that the application cannot change.
+For example, mapping a property containing a list of numbers: 
+
+[!code-csharp[ListProperty](../../../samples/core/Modeling/ValueConversions/MappingListProperty.cs?name=ListProperty)]
+
+The [`List<T>` class](/dotnet/api/system.collections.generic.list-1?view=netstandard-2.1):
+* Has reference equality; two lists containing the same values are treated as different.
+* Is mutable; values in the list can be added and removed.
+
+A typical value conversion on a list property might convert the list to and from JSON:
+
+[!code-csharp[ConfigureListProperty](../../../samples/core/Modeling/ValueConversions/MappingListProperty.cs?name=ConfigureListProperty)]
+
+This then requires setting a `ValueComparer<T>` on the property to force EF Core use correct comparisons with this conversion:
+
+[!code-csharp[ConfigureListPropertyComparer](../../../samples/core/Modeling/ValueConversions/MappingListProperty.cs?name=ConfigureListPropertyComparer)]
+
+> [!NOTE]  
+> The model builder ("fluent") API to set a value comparer has not yet been implemented.
+> Instead, the code above calls SetValueComparer on the lower-level IMutableProperty exposed by the builder as 'Metadata'.
+
+The `ValueComparer<T>` constructor accepts three expressions:
+* An expression for checking quality
+* An expression for generating a hash code
+* An expression to snapshot a value  
+
+In this case the comparision is done by checking if the sequences of numbers are the same.
+
+Likewise, the hash code is built from this same sequence.
+(Note that this is a hash code over mutable values and hence can [cause problems](https://ericlippert.com/2011/02/28/guidelines-and-rules-for-gethashcode/).
+Be immutable instead if you can.)
+
+The snapshot is created by cloning the list with ToList.
+Again, this is only needed if the lists are going to be mutated.
+Be immutable instead if you can. 
+
+> Point of interest:  
+> Value converters and comparers are constructed using expressions rather than simple delegates.
+> This is because EF inserts these expressions into a much more complex expression tree that is then compiled into an entity shaper delegate.
+> Conceptually, this is similar to compiler inlining.
+> For example, a simple conversion may just be a compiled in cast, rather than a call to another method to do the conversion.    
+
+### Key comparers
+
+The background section covers why key comparisons may require special semantics.
+Make sure to create a comparer that is appropriate for keys when setting it on a primary, principal, or foreign key property.
+
+Use [SetKeyValueComparer](/dotnet/api/microsoft.entityframeworkcore.mutablepropertyextensions.setkeyvaluecomparer?view=efcore-3.1) in the rare cases where different semantics is required on the same property.
+
+> [!NOTE]  
+> SetStructuralComparer has been obsoleted in EF Core 5.0.
```

Comment:
`SetStructuralComparer` is totally out of place. Either describe about it, remove this altogether. You can also use previous versions section to add something for past releases.

Reply:
How will that help anyone. All we need here is a note saying that if you were using it, then don't.

Reply:
Obsolete message from the method should tell the replacement in which case we can skip it here altogether. Or we want to keep it here that we should at least write that `SetStructuralComparer` was being used for something now you should use something else.
For a someone who is new to this topic, it just confuses what is used for.